File: 2021/day_1/day_1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"bufio"
);

func main() {
	path := "./day_1.txt"
	arr := readNumbers(path)
	fmt.Println("--- Part One ---")
	fmt.Println(f(arr))

	fmt.Println("--- Part Two ---")
	fmt.Println(f2(arr))
}


func f(arr []int) int {
	res := 0
	for i := 1; i < len(arr); i ++ {
		if arr[i] - arr[i - 1] > 0 {
			res += 1
		}
	}
	return res
}


func f2(arr []int) int {
	res := 0
	prev := 0
	s := 0
	i := 0
	size := 3

	for j, val := range arr {
		s += val
		if j - i + 1 > size {
			s -= arr[i]
			i += 1
		}

		if j - i + 1 == size {
			if s > prev && j != size - 1 {
				res += 1
			}
			prev = s
		}

	}
	return res
}


func readNumbers(filename string) []int {
	file, err := os.Open(filename)
	check(err)
	defer file.Close()

	Scanner := bufio.NewScanner(file)

	var numbers []int
	for Scanner.Scan() {
		numbers = append(numbers, toInt(Scanner.Text()))
	}
	return numbers
}


func check(err error) {
	if err != nil {
		panic(err)
	}
}


func toInt(str string) int {
	res, err := strconv.Atoi(str)
	check(err)
	return res
}

File: 2021/day_1/day_1.py
def f(A):
	n = len(A)
	res = 0
	for i in range(1, n):
		if A[i] - A[i - 1] > 0:
			res += 1
	return res

with open('2021/day_1.txt') as file:
    data = file.read()

A = [int(n) for n in data.split('\n')]
# print(f(A))


def f2(A):
	size = 3
	total = 0
	i = 0
	prev = 0
	res = 0
	for j in range(len(A)):
		total += A[j]
		if j - i + 1 > size:
			total -= A[i]
			i += 1 
		
		if j - i + 1 == size:
			if j == 2:
				prev = total
				continue
			if total > prev:
				res += 1
			prev = total
			
	return res

print(f2(A))
		